doc/source/guide/in_depth/functional_guide.rst

```diff
@@ -0,0 +1,257 @@
+| ``(f * v)(x)``     | ``f(v * x)``    | `FunctionalRightVectorMult`                     |
+|                    |                 | - Retains gradient and convex conjugate.        |
++--------------------+-----------------+-------------------------------------------------+
+| ``f.translate(y)`` | ``f(. - y)``    | `TranslatedFunctional`                          |
+|                    |                 | - Retains all properties.                       |
++--------------------+-----------------+-------------------------------------------------+
+
+
+Code example
+============
+This section contains an example of an implementation of a functional, namely the functional :math:`\|x\|_2^2 + \langle x, y \rangle`.
+Another example can be found `functional_basic_example.py`, and more implementations of other functionals can be found in `default_functionals.py`. ::
+
+    # Here we define the functional
+    class MyFunctional(odl.solvers.Functional):
+        """This is my functional: ||x||_2^2 + <x, y>."""
+
+        def __init__(self, domain, y):
+            # This comand calls the init of Functional and sets a number of
+            # parameters associated with a functional. All but domain have default
+            # values if not set.
+            super().__init__(domain=domain, linear=False, convex=True,
+                             concave=False, smooth=True, grad_lipschitz=2)
+
+            # We need to check that y is in the domain. Then we store the value of
+            # y for future use.
+            if y not in domain:
+                raise TypeError('y is not in the domain!')
+            self._y = y
+
+        # Property that returns the linear term.
+        @property
+        def y(self):
+            return self._y
+
+        # Defining the _call function
+        def _call(self, x):
+            return x.norm()**2 + x.inner(self.y)
+
+        # Next we define the gradient. Note that this is a property.
+        @property
+        def gradient(self):
+
+            # The class corresponding to the gradient operator.
+            class MyGradientOperator(odl.Operator):
+                """Class that implements the gradient operator of the functional
+                ``||x||_2^2 + <x,y>``.
+                """
+
+                def __init__(self, functional):
+                    super().__init__(domain=functional.domain,
+                                     range=functional.domain)
+
+                    self._functional = functional
+
+                def _call(self, x):
+                    return 2.0 * x + self._functional.y
+
+            return MyGradientOperator(functional=self)
+
+        # Next we define the convex conjugate functional.
+        @property
+        def conjugate_functional(self):
+            # This functional is implemented below.
+            return MyFunctionalConjugate(domain=self.domain, y=self.y)
+
+
+    # Here is the conjugate functional.
```

Comment:
I was avoiding to many inner class structures, in order to make it simpler. I think I'll keep it with two classes, if you don't strongly advocate the inner-class construction.